### ISSUE TITLE
Distinguishing `float`s and `int`s when writing to Yaml

### ISSUE THREAD

**Issue by ARessegetesStery** [fktn-k/fkYAML#405]
### Description

Writing `float` to Yaml will not have `.0` suffix and thus cannot be read by `float` again using the library.

### Reproduction steps

Try
```
fkyaml::node n = 0.0;             // double
std::stringstream ss;
ss << n;
ss.str();                                   // will give 0 instead of 0.0

// I haven't tried to deserialize a stringstream but suppose that this is an ifstream
fkyaml::node test = fkyaml::node::deserialize(ss);    
test.get_value<double>();     // gives exception: type mismatch (int vs. double)
```

### Expected vs. actual results

The above code should work as it is reasonable to get another double if a double is written to the yaml.

### Compiler and operating system

g++; Windows

### Library version

v0.3.12

### Validation

- [X] The bug also occurs if the latest version from the [`develop`](https://github.com/fktn-k/fkYAML/tree/develop) branch is used.
- [ ] I can successfully [compile and run the unit tests](https://github.com/fktn-k/fkYAML#how-to-execute-the-unit-tests).

**Comment by ARessegetesStery** [comment]
It may be helpful to do the following changes (definitely there should be a better choice, but I haven't though of any)

Modify the overloaded operator `<<` for `FloatType` (for header only version, the lines 9549-9574) from
```
std::ostringstream oss;
oss << f;
s = oss.str();
```
to
```
std::ostringstream oss;
oss.setf(std::ios::fixed);
oss << std::setprecision(6) << f;
s = oss.str();
```
It may not be a good idea to fix precision beforehand for all applications, but at least this will ensure that writing a `double` to yaml and reading that as a `double` afterwards will not give any exceptions

**Comment by fktn-k** [comment]
@ARessegetesStery   
Thanks for filing the issue and proposing a fix!  
The false serialization from `0.0` to `"0"` happens in my environment too.  

(FYI, I also found an issue in deserializing a YAML document which contains only one scalar, but that's not relevant to this issue. I'll fix it later as another issue.)  

Regarding your suggestion, I agree to fix precision before calling operator << to resolve the roundtrip issue. I can't come up with any other using `std::ostringstream` neither. Although I honestly want to stop using it due to its performance/locale issues, fixed precision should be enough for now. But I want to trim trailing zeros before writing to an output buffer. They often make serialization outputs ugly...

**Comment by ARessegetesStery** [comment]
It may be possible to do a special judgement for `FloatType`s which are actually integers
```
FloatType f;
std::stringstream ss;
ss << f;
std::string str = ss.str();
if (f - std::floor(f) < std::numeric_limits<FloatType>()::lowest())
    str = str + ".0";
```
Or, as an alternative approach, trim the trailng zeros after obtaining the string
```
FloatType f;
std::stringstream ss;
ss << f;
std::string str = ss.str();
str = str.substr(0, str.find_last_not_of('0')+1);
if (str.find('.') == str.size() - 1)
    str = str + "0";
```

**Comment by fktn-k** [comment]
@ARessegetesStery   
I like the first alternative better, i.e., a special judgement for FloatTypes which are actually integers since there's no need for tricky string manipulations.  
I tried the snippet, but it doesn't work as expected. `std::numeric_limits<Float>()::lowest()` should be changed to `std::numeric_limits<FloatType>::min()` ? The `min()` version successfully appends ".0" whether `f` is either positive or negative.  

```cpp
FloatType f;
std::stringstream ss;
ss << f;
std::string str = ss.str();
if (f - std::floor(f) < std::numeric_limits<FloatType>()::min()) // lowest() -> min()
    str = str + ".0";
```

**Comment by ARessegetesStery** [comment]
@fktn-k 
I guess you are right -- I am not so familiar with std::numeric_limits lol
Thanks for fixing this!

**Comment by fktn-k** [comment]
@ARessegetesStery   
Are you fixing this issue? If not, I can do it.

**Comment by ARessegetesStery** [comment]
@fktn-k 
I am not.. Please do it. Thanks!
(I guess both the header-only version and the full version need to be changed -- and I am using the header-only version, without having a look at the full version lol)

**Comment by fktn-k** [comment]
@ARessegetesStery   
OK.  I'll do it and let you know when the bug is fixed.  

FYI, the header-only version is just an amalgamation of the “full“ version's source files, so you just need to change and amalgamate them.

**Comment by fktn-k** [comment]
@ARessegetesStery   
(Reopened since GitHub automatically closed this issue before your confirmation.)  

This issue should be fixed by the PR #407.  
Can you confirm with the develop HEAD?

**Comment by ARessegetesStery** [comment]
@fktn-k 
I've tried running the code and it seems to function as expected. This should be ready for merging
Thanks for checking!

**Comment by fktn-k** [comment]
@ARessegetesStery   
Thanks for your confirmation!  
The fix is merged to the develop branch.  

Do you need it to be released soon? It would otherwise be released in a week or two with some more changes.

**Comment by ARessegetesStery** [comment]
@fktn-k 
No need to hurry -- feel free to fit this change into your release schedules :)
Thanks again for the timely response and fix!

**Comment by fktn-k** [comment]
@ARessegetesStery   
OK. Thanks!  
I'll close this issue as resolved.  
Feel free to reopen it if you noticed anything relevant but still unresolved.

**Comment by dyerbod** [comment]
The solution doesn't work for numbers that get serialised using scientific notation. See snip taken while debugging. Adding ".0" in such cases results in the output being parsed as a string.
![image](https://github.com/user-attachments/assets/070fcfd6-6511-4c03-a389-769a42808ac9)

**Comment by fktn-k** [comment]
@dyerbod   
You're right. We should check if scientific notation is used before adding “.0“.  
I'll fix it this weekend. Thanks!

**Comment by fktn-k** [comment]
@dyerbod  
I created a pull request #439 to fix the reported issue.  
Does that actually resolve your problem?

**Comment by dyerbod** [comment]
Yes, thank you. When might the fix be incorporated into a release tag?

**Comment by fktn-k** [comment]
@dyerbod   
Thanks for your confirmation:)  
I just merged the fix into the develop branch.  


> When might the fix be incorporated into a release tag?

I'm publishing a release within a few days if no critical issue is found.

**Comment by fktn-k** [comment]
Close this issue as the reported issue is resolved. (The fix has been released as v0.4.0)  
Feel free to reopen it if any related issue is found.
